Sources/Flatten.swift

```diff
@@ -288,18 +288,13 @@ extension Signal where Value: SignalProducerConvertible, Error == Value.Error {
 	fileprivate func concurrent(limit: UInt) -> Signal<Value.Value, Error> {
 		precondition(limit > 0, "The concurrent limit must be greater than zero.")
 
-		return Signal<Value.Value, Error> { relayObserver in
-			let disposable = CompositeDisposable()
-			let relayDisposable = CompositeDisposable()
-
-			disposable += relayDisposable
-			disposable += self.observeConcurrent(relayObserver, limit, relayDisposable)
-
-			return disposable
+		return Signal<Value.Value, Error> { relayObserver, lifetime in
+			let disposable = self.observeConcurrent(relayObserver, limit, lifetime)
+			_ = (disposable?.dispose).map(lifetime.observeEnded)
```

Comment:
I strongly dislike using `map` like this. `map` shouldn't be used for side effects. This is also not very clear IMO.

I think we're better off with `if let dispose = disposable?.dispose { lifetime.observeEnded(dispose) }`.

Reply:
`lifetime += disposable` should work, and I have already replaced most of them.